src/processor/tree/mod.rs

```diff
@@ -26,30 +26,31 @@ pub struct TreeProcessor {
     /// The internal merkle tree.
     tree: TreeWrapper,
     /// The stored state snapshot.
-    snapshot: Arc<Mutex<StateSnapshot>>,
+    snapshot: Arc<Mutex<InnerDB>>,
 }
 
 impl TreeProcessor {
-    pub async fn new(db_path: PathBuf, snapshot: Arc<Mutex<StateSnapshot>>) -> Result<Self> {
-        // If database directory already exists, we try to restore the latest state.
-        // The state contains the last processed block and a mapping of index to key
-        // values, if a state file does not exist, we simply use the defaults instead.
-        let should_restore_state = db_path.exists();
-        if should_restore_state {
-            tracing::info!("Loading previous state file...");
-            let new_state = StateSnapshot::read(&db_path.join(STATE_FILE_NAME))
-                .expect("state file is malformed");
-            *snapshot.lock().await = new_state;
-        } else {
-            tracing::info!("No existing database found, starting from genesis...");
-        };
+    pub async fn new(db_path: PathBuf) -> Result<Self> {
+        // If database directory already exists, we try to restore the
+        // latest state from a database inside of it.  The state
+        // contains the last processed block and a mapping of index to
+        // key values.
+        let inner_db_path = db_path.join(INNER_DB_NAME);
+        let init = !db_path.exists() || !inner_db_path.exists();
```

Comment:
How do we handle the case where `db_path` exists but not `inner_db_path`? If we initialize with the genesis state, when we already have had some sync progress, I believe the state tree database would get corrupted.

Notably, the case where `db_path` doesn't exist but `inner_db_path` isn't possible, so perhaps we should panic and give a nice message that a critical part of the database is missing and as such we cannot allow that database instance to sync further? E.g. first make sure that the state db exists, and only if it does do we check that the inner one does too.

Reply:
OK, let's just disallow the case.